runtime/src/main/java/org/corfudb/runtime/collections/streaming/DeltaStream.java

```diff
@@ -0,0 +1,184 @@
+            // to read, notice that a trim exception can occur even if streamaddressspace is empty
+            trimMark = streamAddressSpace.getTrimMark();
+
+            // if the task is syncing it will either discover the trim during reading the addresses that
+            // are already buffered, or discover this trim the next time the task gets scheduled
+
+            // since refresh can be called while the task is syncing we can error out when the scheduling tries to
+            // reschedule this task
+            return;
+        }
+
+        Preconditions.checkState(trimMark <= streamAddressSpace.getTrimMark(),
+                "streamId=%s new trimMark %s has to be >= %s", streamId, streamAddressSpace.getTrimMark(), trimMark);
+
+        // Adopt new trim mark
+        trimMark = streamAddressSpace.getTrimMark();
+
+        int maxAddressesToTransfer = availableSpace();
+        long[] newAddresses = streamAddressSpace.toArray();
+        for (int idx = 0; idx < newAddresses.length && idx < maxAddressesToTransfer; idx++) {
+            // if maxAddressesToTransfer doesn't fit, then the rest of the addresses will be
+            // thrown away and re-discovered on the next pool
+            // Note that this predicate can fail in certain edge cases where the sequencer
+            // can regress. This is an existing but in the current streaming layer and this
+            // PR doesn't attempt to fix the issue, but rather fail explicitly instead of
+            // ignoring the fault.
+            Preconditions.checkState(maxAddressSeen < newAddresses[idx],
+                    "maxAddressSeen %s not < %s", maxAddressSeen, newAddresses[idx]);
+            Preconditions.checkState(addressesToRead.add(newAddresses[idx]),
+                    "failed to queue %d", newAddresses[idx]);
+            // Only update max seen after it has been added to the buffer
+            maxAddressSeen = newAddresses[idx];
+            log.trace("refresh: transferring {} id {}", newAddresses, streamId);
+        }
+    }
+
+    /**
+     * Has next indicates whether there is more data to read, or if there has been a trimmed exception detected.
+     * @return return true if a there are pending reads, or if a trimmed exception occurred, otherwise return false
+     */
+    public boolean hasNext() {
+        return !addressesToRead.isEmpty() || trimGap();
+    }
+
+    public ILogData next() {
+
+        if (lastAddressRead < trimMark) {
+            throw new TrimmedException(String.format("lastAddressRead %d trimMark %d", lastAddressRead, trimMark));
+        }
+
+        long readAddress = (long) addressesToRead.remove();
+
+        Preconditions.checkState(lastAddressRead < readAddress,
+                "%d must be greater than %d", readAddress, lastAddressRead);
+
+        if (readAddress <= trimMark) {
+            throw new TrimmedException(String.format("current %d trimMark %d", readAddress, trimMark));
```

Comment:
Do we have metrics around how much time an address sits on the buffer before being serviced?

Reply:
I added a queuing delay metric per stream.